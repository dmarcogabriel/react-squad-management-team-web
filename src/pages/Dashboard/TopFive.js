import React from 'react';
import Card from '../../common/Card';

const TopFive = () => {
  return (
    <Card title="Top 5">
      <div className="flex flex-col md:flex-row mx-2">
        <div className="w-full mb-5 md:mr-8">
          <h2 className="text-base mb-3">Highest avg age</h2>

          <div className="bg-background rounded-lg p-1">
            {Array.from(Array(5)).map((_, i) => (
              <div
                key={String(i)}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p className="text-sm">Arsenal FC</p>
                <p className="font-bold text-sm">31.2</p>
              </div>
            ))}
          </div>
        </div>

        <div className="w-full">
          <h2 className="text-base mb-3">Lowest avg age</h2>

          <div className="bg-background rounded-lg p-1">
            {Array.from(Array(5)).map((_, i) => (
              <div
                key={String(i)}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p className="text-sm">Arsenal FC</p>
                <p className="font-bold text-sm">31.2</p>
              </div>
            ))}
          </div>
        </div>
      </div>
    </Card>
  );
};

export default TopFive;
