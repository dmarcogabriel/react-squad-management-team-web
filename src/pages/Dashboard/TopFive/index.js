import React, { useEffect, useState } from 'react';
import Card from '../../../common/Card';
import { useSquad } from '../../../contexts/Squad';

const TopFive = () => {
  const [highestAvgAge, setHighestAvgAge] = useState([]);
  const [lowestAvgAge, setLowestAvgAge] = useState([]);

  const [squads] = useSquad();

  const getPlayersAverage = () => {
    const lowestAvg = squads
      .sort((sq1, sq2) => sq1.ageAvg - sq2.ageAvg)
      .slice(0, 5);

    const highestAvg = squads
      .sort((sq1, sq2) => sq1.ageAvg - sq2.ageAvg)
      .reverse()
      .slice(0, 5);

    setHighestAvgAge(highestAvg);
    setLowestAvgAge(lowestAvg);
  };

  useEffect(() => {
    getPlayersAverage();
  }, []);

  return (
    <Card title="Top 5">
      <div className="flex flex-col md:flex-row mx-2">
        <div className="w-full mb-5 md:mr-8">
          <h2 className="text-base mb-3">Highest avg age</h2>

          <div
            data-testid="highest-avgs"
            className="bg-background rounded-lg p-1"
          >
            {highestAvgAge.map((squad) => (
              <div
                key={squad.id}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p data-testid={`highest-${squad.id}-name`} className="text-sm">
                  {squad.name}
                </p>
                <p className="font-bold text-sm">{squad.ageAvg.toFixed(1)}</p>
              </div>
            ))}
          </div>
        </div>

        <div className="w-full">
          <h2 className="text-base mb-3">Lowest avg age</h2>

          <div
            data-testid="lowest-avgs"
            className="bg-background rounded-lg p-1"
          >
            {lowestAvgAge.map((squad) => (
              <div
                key={squad.id}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p data-testid={`lowest-${squad.id}-name`} className="text-sm">
                  {squad.name}
                </p>
                <p className="font-bold text-sm">{squad.ageAvg.toFixed(1)}</p>
              </div>
            ))}
          </div>
        </div>
      </div>
    </Card>
  );
};

export default TopFive;
