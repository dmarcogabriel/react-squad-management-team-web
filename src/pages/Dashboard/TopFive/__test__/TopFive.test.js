import { cleanup, render } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TopFive from '..';

jest.mock('contexts/Squad', () => ({
  useSquad: () => [
    [
      {
        id: '1',
        ageAvg: 10,
        name: 'Test Highest',
      },
      {
        id: '2',
        ageAvg: 9,
        name: 'Test Middle',
      },
      {
        id: '3',
        ageAvg: 8,
        name: 'Test Middle',
      },
      {
        id: '4',
        ageAvg: 7,
        name: 'Test Middle',
      },
      {
        id: '5',
        ageAvg: 6,
        name: 'Test Lowest',
      },
    ],
  ],
}));

describe('<TopFive />', () => {
  afterEach(cleanup);

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TopFive />, div);
  });

  it('renders highest and lowest avg squads', () => {
    const { getByTestId, rerender } = render(<TopFive />);

    rerender(<TopFive />);

    const highestAvg = getByTestId('highest-avgs');
    const lowestAvg = getByTestId('lowest-avgs');

    expect(highestAvg.childElementCount).toBe(5);
    expect(lowestAvg.childElementCount).toBe(5);
  });

  it('render squad "Test Highest" first', () => {
    const { getByTestId, rerender } = render(<TopFive />);

    rerender(<TopFive />);

    const lowest = getByTestId('lowest-5-name');
    const highestSquad = getByTestId('highest-1-name');

    expect(highestSquad).toHaveTextContent('Test Highest');
    expect(lowest).toHaveTextContent('Test Lowest');
  });

  it('render squad "Test Lowest" first and "Test Highest" last', () => {
    const { getByTestId, rerender } = render(<TopFive />);

    rerender(<TopFive />);

    const lowest = getByTestId('lowest-1-name');
    const highest = getByTestId('highest-5-name');

    expect(highest).toHaveTextContent('Test Lowest');
    expect(lowest).toHaveTextContent('Test Highest');
  });

  it('match snapshot', () => {
    expect(renderer.create(<TopFive />).toJSON()).toMatchSnapshot();
  });
});
