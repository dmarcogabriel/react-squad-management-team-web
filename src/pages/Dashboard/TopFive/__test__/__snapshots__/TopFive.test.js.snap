// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopFive /> match snapshot 1`] = `
<div
  className="bg-white rounded-lg shadow-lg my-10 w-full"
  data-testid="card"
>
  <div
    className="flex justify-between items-center p-8 border-b-2 border-gray-200"
  >
    <h1
      className="text-secondary-dark font-bold text-md"
      data-testid="card-title"
    >
      Top 5
    </h1>
  </div>
  <div
    className="py-5 px-2"
    data-testid="card-children"
  >
    <div
      className="flex flex-col md:flex-row mx-2"
    >
      <div
        className="w-full mb-5 md:mr-8"
      >
        <h2
          className="text-base mb-3"
        >
          Highest avg age
        </h2>
        <div
          className="bg-background rounded-lg p-1"
          data-testid="highest-avgs"
        />
      </div>
      <div
        className="w-full"
      >
        <h2
          className="text-base mb-3"
        >
          Lowest avg age
        </h2>
        <div
          className="bg-background rounded-lg p-1"
          data-testid="lowest-avgs"
        />
      </div>
    </div>
  </div>
</div>
`;
